### ISSUE TITLE
GeneName and GeneID fail

### ISSUE THREAD

**Issue by kokyriakidis** [hbc/bcbioRNASeq#171]
Hi @mjsteinbaugh 

In recent runs GeneName and GeneID fail to be annotated. Have you observed this behavior? Plots get "Name11124" kind of names

**Comment by mjsteinbaugh** [comment]
Hi @kokyriakidis, can you post an example and your R `sessionInfo()` when bcbioRNASeq is loaded?

I'll take a look and troubleshoot this week.

**Comment by kokyriakidis** [comment]
> colnames(rowData(object))
[1] "broadClass"    "geneBiotype"   "geneId"        "geneIdVersion" "geneName"      "geneSource"   
[7] "source"        "type"

**Comment by mjsteinbaugh** [comment]
And the metadata there looks OK correct? It's just a plotting issue? Which plot function is not working as expected?

**Comment by kokyriakidis** [comment]
geneName under elementMetadata shows gene names correctly.

trying to plot with plotMA gets me
```
Error in (function (classes, fdef, mtable) : unable to find an inherited method for function ‘mapGenesToRownames’ for signature ‘"Gene2Symbol"’
```

Then `resultsTables(object_da, i = 1L)` ouputs a table with no geneID and geneName (its like id9599 and name9599)

**Comment by kokyriakidis** [comment]
I have also another issue:

When I try to get the results with the following commands:

```
res_list_unshrunken <- results(dds, alpha = params$alpha, lfcThreshold = params$lfc_threshold, contrast=params$contrasts[[1]])

assignAndSaveData(
    name = paste(name, "res_list_unshrunken", sep = "_"),
    object = res_list_unshrunken,
    dir = params$data_dir
)
```
```
> params$contrasts[[1]]
     factor   numerator denominator 
 "category"        "AD"   "Control" 
```
It runs OK. Then I try to calculate the shrunken log2fc with "apeglm" or "normal" with these commands:

```
resultsNames(dds)

res_list_shrunken <- lfcShrink(dds, type="apeglm", coef = 2, res = res_list_unshrunken)


assignAndSaveData(
    name = paste(name, "res_list_shrunken", sep = "_"),
    object = res_list_shrunken,
    dir = params$data_dir
)
```
and I get the following error:
```
[1] "Intercept"              "category_Control_vs_AD"
Error in DESeq2::lfcShrink(object, ...) : 'coef' should specify same coefficient as in results 'res'
```
The first time I run the Rmd analysis it worked. When I tried to run it again it failed in this step.

**Comment by mjsteinbaugh** [comment]
Perfect thanks, I know how to fix the gene mapping issue and will patch that.

I'll take a look to see if I can reprex the DESeq2 `lfcShrink` issue. For `lfcShrink` with the `normal` shrinkage method, I recommend using the `contrast` argument (see `DESeq::lfcShrink` for details).

Here's the current default parameterized template code, for reference:

```r
res_list_shrunken <- mapply(
    FUN = lfcShrink,
    res = res_list_unshrunken,
    contrast = contrasts,
    MoreArgs = list(
        dds = dds,
        type = "normal",
        alpha = params$alpha,
        lfcThreshold = params$lfc_threshold
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
)
assignAndSaveData(
    name = paste(name, "res_list_shrunken", sep = "_"),
    object = res_list_shrunken,
    dir = params$data_dir
)
```

The `apeglm` method only currently supports `coef`, which can be a bit tricky to parameterize correctly. I wrote a function named [`apeglmResults`](https://r.acidgenomics.com/packages/deseqanalysis/reference/apeglmResults.html) that helps add support for the `contrast` argument, if you want to take a look and try it.

You may be hitting the error because of `res = res_list_unshrunken`, which seems to be attempting to pass in a `list` instead of a `DESeqResults` object for the `res` argument.

Best,
Mike

**Comment by kokyriakidis** [comment]
It does not work:
```
> res_list_shrunken <- mapply(
+     FUN = lfcShrink,
+     res = res_list_unshrunken,
+     contrast = contrasts,
+     MoreArgs = list(
+         dds = dds,
+         type = "normal",
+         alpha = params$alpha,
+         lfcThreshold = params$lfc_threshold
+     ),
+     SIMPLIFY = FALSE,
+     USE.NAMES = TRUE
+ )
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘lfcShrink’ for signature ‘"missing"’
> assignAndSaveData(
+     name = paste(name, "res_list_shrunken", sep = "_"),
+     object = res_list_shrunken,
+     dir = params$data_dir
+ )
Error in assert(isString(name), !is.null(object), formalCompress(compress),  : 
  object 'res_list_shrunken' not found
```

**Comment by kokyriakidis** [comment]
Yes this step works fine. The next step does not work (the one that creates the res_list_unshrunken file). It gives me:
```
Error in (function (classes, fdef, mtable) : 
  unable to find an inherited method for function ‘lfcShrink’ for signature ‘"missing"’
```
I run the whole template fine until this step. 

res_list_unshrunken is a Large list with Formal class 'DESeqResults'

When I just try to run `res_list_shrunken <- lfcShrink(res_list_unshrunken)` I get the following error:
```
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘lfcShrink’ for signature ‘"list"’
```
DESeqResults do not work either:
```
> lfcShrink(res_list_unshrunken[[1]])
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘lfcShrink’ for signature ‘"DESeqResults"’
```
Should there be a method for DESeqResults? 
```
> showMethods(lfcShrink) 
Function: lfcShrink (package AcidGenerics)
object="DESeqAnalysis"
object="DESeqAnalysisList"
object="DESeqDataSet"
```
Running `res <- lfcShrink(dds, coef=2)` works fine because it is a "DESeqDataSet".

There must be something wrong with how it handles `res_list_unshrunken` "DESeqResults" format.

In order to make `apeglmResults` to work I had to write it in the following way:
```
res_list_shrunken <- apeglmResults(
        object = dds,
        contrast = contrasts[[1]],
        res = res_list_unshrunken[[1]])
``` 
`lfcShrinkType(res_list_unshrunken[[1]])` should be 'unshrunken' and not a list.

**Comment by kokyriakidis** [comment]
I continued with the unshrunken to see how this goes and got an error at the export stage:
```{r results-tables}
## Here we're creating subset tables of the DEGs, and adding the normalized
## counts used by DESeq2 for the differential expression analysis.
res_tbl_list <- mapply(
    FUN = resultsTables,
    object = res_list_unshrunken,
    MoreArgs = list(
        DESeqDataSet = dds,
        alpha = params$alpha,
        lfcThreshold = params$lfc_threshold
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
)
assignAndSaveData(
    name = paste(name, "res_tbl_list", sep = "_"),
    object = res_tbl_list,
    dir = params$data_dir
)
```
When I got the following error:
```
Error in .local(object, ...) : unused argument (DESeqDataSet = new("DESeqDataSet", design = ~category, dispersionFunction = function (q) coefs[1] + coefs[2]/q, rowRanges = new("GRanges", seqnames = new("Rle", values = c(23, 20, 1, 6, 1, 23, 6, 3, 7, 12, 7, 11, 4, 23, 4, 7, 17, 7, 12, 23, 2, 7, 16, 2, 3, 8, 7, 17, 3, 1, 4, 12, 1, 3, 17, 12, 7, 19, 16, 7, 6, 3, 7, 23, 7, 17, 7, 23, 17, 19, 7, 23, 4, 7, 17, 7, 17, 12, 16, 19, 9, 17, 7, 23, 16, 17, 7, 2, 7, 16, 6, 11, 13, 16, 17, 23, 17, 7, 16, 19, 17, 11, 16, 11, 17, 7, 16, 17, 16, 7, 14, 7, 11, 1, 7, 2, 11, 7, 19, 7, 17, 7, 19, 7, 17, 7, 23, 17, 12, 5, 2, 19, 4, 19, 17, 19, 17, 1, 23, 11, 16, 3, 16, 6, 1, 6, 23, 1, 6, 23, 6, 23, 1, 6, 3, 7, 17, 3, 7, 3, 19, 12, 19, 8, 16, 22, 17, 8, 2, 3, 8, 1, 7, 6, 23, 1, 6, 1, 14, 6, 7, 6, 1, 12, 17, 3, 7, 12, 3, 12, 19, 3, 19, 12, 23, 9, 16, 12, 23, 6, 1, 6, 1, 17, 5, 12, 14, 19, 17, 3, 23, 19, 17, 7, 17, 19, 11, 3, 11, 19, 7, 19, 9, 12, 19, 2, 19, 16, 23, 17, 19, 3, 23, 3, 8, 12, 6, 10, 24, 12, 14, 2, 19, 4, 3, 17, 16,
8.
(new("nonstandardGenericFunction", .Data = function (object, ...) { standardGeneric("resultsTables") ...
7.
(new("nonstandardGenericFunction", .Data = function (object, ...) { standardGeneric("resultsTables") ...
6.
mapply(FUN = FUN, ..., MoreArgs = MoreArgs, SIMPLIFY = SIMPLIFY, USE.NAMES = USE.NAMES)
5.
eval(mc, env)
4.
eval(mc, env)
3.
eval(mc, env)
2.
standardGeneric("mapply")
1.
mapply(FUN = resultsTables, object = res_list_unshrunken, MoreArgs = list(DESeqDataSet = dds, alpha = params$alpha, lfcThreshold = params$lfc_threshold), SIMPLIFY = FALSE, USE.NAMES = TRUE)
```

**Comment by kokyriakidis** [comment]
I have a working version of DE analysis. I will sent it to you once I correct some things.

I have one issue. When I plot figs inside RStudio I get the the correct contrast eg. category_AD_vs_Controls`. When I knit the document I get the opposite contrast eg. `category_Control_vs_AD`

Do you have any idea why this is happening and how to resolve it? Results inside the graph are also flipped!

**Comment by mjsteinbaugh** [comment]
OK great thanks, I know how to fix a couple of these issues and will post an update soon.

**Comment by mjsteinbaugh** [comment]
@kokyriakidis OK I'm working on patching this in the package today

**Comment by mjsteinbaugh** [comment]
OK here's a minimal repex -- let me know if this works OK for you. Otherwise there may be a dependency package that needs to be updated.

```r
suppressPackageStartupMessages({
    library(bcbioRNASeq)
    library(DESeq2)
    library(DESeqAnalysis)
})

sessionInfo()
## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Big Sur 11.4
##
## Matrix products: default
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
##
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
##
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base
##
## other attached packages:
## [1] bcbioRNASeq_0.3.40  basejump_0.14.19    AcidDevTools_0.3.16
##
## loaded via a namespace (and not attached):
##  [1] MatrixGenerics_1.4.0        Biobase_2.52.0
##  [3] httr_1.4.2                  edgeR_3.34.0
##  [5] splines_4.1.0               bit64_4.0.5
##  [7] AcidMarkdown_0.1.1          assertthat_0.2.1
##  [9] stats4_4.1.0                blob_1.2.1
## [11] GenomeInfoDbData_1.2.6      sessioninfo_1.1.1
## [13] pillar_1.6.1                RSQLite_2.2.7
## [15] lattice_0.20-44             limma_3.48.1
## [17] glue_1.4.2                  RColorBrewer_1.1-2
## [19] AcidGenomes_0.2.14          GenomicRanges_1.44.0
## [21] XVector_0.32.0              AcidSingleCell_0.1.7
## [23] colorspace_2.0-2            cowplot_1.1.1
## [25] Matrix_1.3-4                DESeq2_1.32.0
## [27] XML_3.99-0.6                pkgconfig_2.0.3
## [29] AcidPlyr_0.1.20             genefilter_1.74.0
## [31] zlibbioc_1.38.0             AcidCLI_0.1.2
## [33] purrr_0.3.4                 xtable_1.8-4
## [35] scales_1.1.1                bcbioBase_0.6.21
## [37] goalie_0.5.2                BiocParallel_1.26.1
## [39] tibble_3.1.2                annotate_1.70.0
## [41] KEGGREST_1.32.0             generics_0.1.0
## [43] IRanges_2.26.0              ggplot2_3.3.5
## [45] AcidPlots_0.3.7             ellipsis_0.3.2
## [47] cachem_1.0.5                withr_2.4.2
## [49] SummarizedExperiment_1.22.0 AcidExperiment_0.1.12
## [51] AcidGenerics_0.5.18         BiocGenerics_0.38.0
## [53] cli_3.0.1                   survival_3.2-11
## [55] magrittr_2.0.1              crayon_1.4.1
## [57] memoise_2.0.0               fansi_0.5.0
## [59] syntactic_0.4.5             pipette_0.6.2
## [61] tools_4.1.0                 data.table_1.14.0
## [63] lifecycle_1.0.0             matrixStats_0.59.0
## [65] stringr_1.4.0               S4Vectors_0.30.0
## [67] locfit_1.5-9.4              munsell_0.5.0
## [69] DelayedArray_0.18.0         Biostrings_2.60.1
## [71] AnnotationDbi_1.54.1        compiler_4.1.0
## [73] GenomeInfoDb_1.28.1         rlang_0.4.11
## [75] grid_4.1.0                  RCurl_1.98-1.3
## [77] tximport_1.20.0             SingleCellExperiment_1.14.1
## [79] bitops_1.0-7                AcidBase_0.3.14
## [81] gtable_0.3.0                DBI_1.1.1
## [83] R6_2.5.0                    knitr_1.33
## [85] dplyr_1.0.7                 fastmap_1.1.0
## [87] bit_4.0.4                   utf8_1.2.1
## [89] stringi_1.7.3               parallel_4.1.0
## [91] Rcpp_1.0.7                  geneplotter_1.70.0
## [93] png_0.1-7                   vctrs_0.3.8
## [95] tidyselect_1.1.1            xfun_0.24

## Load example bcbioRNASeq object from F1000 paper.
loadRemoteData("https://github.com/hbc/bcbioRNASeq/raw/f1000v2/data/bcb.rda")
bcb <- updateObject(bcb)

## Set our P value and log2 fold change cutoffs.
alphaThreshold <- 0.05
lfcThreshold <- 0L

## Run DESeq2.
dds <- as(bcb, "DESeqDataSet")
design(dds) <- ~day
dds <- DESeq(dds)
vst <- varianceStabilizingTransformation(dds)

print(resultsNames(dds))
## [1] "Intercept"  "day_1_vs_0" "day_3_vs_0" "day_7_vs_0"

contrasts <- list(
    "day_1_vs_0" = c(
        "factor" = "day",
        "numerator" = 1L,
        "denominator" = 0L
    ),
    "day_3_vs_0" = c(
        "factor" = "day",
        "numerator" = 3L,
        "denominator" = 0L
    ),
    "day_7_vs_0" = c(
        "factor" = "day",
        "numerator" = 7L,
        "denominator" = 0L
    )
)

res_list_unshrunken <- mapply(
    FUN = DESeq2::results,
    contrast = contrasts,
    MoreArgs = list(
        "object" = dds,
        "alpha" = alphaThreshold,
        "lfcThreshold" = lfcThreshold
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = FALSE
)
names(res_list_unshrunken) <- names(contrasts)

res_list_shrunken <- mapply(
    FUN = DESeq2::lfcShrink,
    res = res_list_unshrunken,
    contrast = contrasts,
    MoreArgs = list(
        "dds" = dds,
        "type" = "normal",
        "alpha" = alphaThreshold,
        "lfcThreshold" = lfcThreshold
    ),
    SIMPLIFY = FALSE,
    USE.NAMES = TRUE
)

## See `help(topic = DESeqAnalysis, DESeqAnalysis)` for details.
## https://r.acidgenomics.com/packages/deseqanalysis/reference/DESeqAnalysis.html
deseq <- DESeqAnalysis(
    data = dds,                     # DESeqDataSet
    transform = vst,                # DESeqTransform
    results = res_list_unshrunken,  # DESeqResults list (unshrunken)
    lfcShrink = res_list_shrunken   # DESeqResults list (shrunken)
)

## Check that MA plots are working as expected.
mapply(
    FUN = plotMA,
    i = resultsNames(deseq),
    MoreArgs = list(object = deseq),
    SIMPLIFY = FALSE
)

## Check results tables support.
resTbl <- resultsTables(deseq, i = 1L)
print(resTbl$all)
## # A tibble: 51,652 x 24
##    rowname   baseMean log2FoldChange  lfcSE   stat    pvalue     padj geneId  
##    <chr>        <dbl>          <dbl>  <dbl>  <dbl>     <dbl>    <dbl> <chr>   
##  1 ENSMUSG0…  1274.            0.861  0.351  2.45    1.42e-2  9.58e-2 ENSMUSG…
##  2 ENSMUSG0…     0            NA     NA     NA      NA       NA       ENSMUSG…
##  3 ENSMUSG0…    23.7           1.42   0.694  2.09    3.65e-2  1.79e-1 ENSMUSG…
##  4 ENSMUSG0…     2.37          1.53   0.759  1.79    7.30e-2 NA       ENSMUSG…
##  5 ENSMUSG0…    18.0          -0.567  0.786 -0.759   4.48e-1  7.40e-1 ENSMUSG…
##  6 ENSMUSG0…    24.2          -0.311  0.718 -0.426   6.70e-1  8.71e-1 ENSMUSG…
##  7 ENSMUSG0…   332.           -1.16   0.485 -2.39    1.68e-2  1.08e-1 ENSMUSG…
##  8 ENSMUSG0…   987.            0.788  0.402  1.96    4.98e-2  2.18e-1 ENSMUSG…
##  9 ENSMUSG0…  2560.            2.44   0.495  4.96    6.92e-7  2.82e-5 ENSMUSG…
## 10 ENSMUSG0…   562.            0.163  0.281  0.578   5.63e-1  8.13e-1 ENSMUSG…
## # … with 51,642 more rows, and 16 more variables: geneName <chr>,
## #   geneBiotype <fct>, description <chr>, broadClass <fct>,
## #   control_rep1 <dbl>, control_rep2 <dbl>, control_rep3 <dbl>,
## #   fa_day1_rep1 <dbl>, fa_day1_rep2 <dbl>, fa_day1_rep3 <dbl>,
## #   fa_day3_rep1 <dbl>, fa_day3_rep2 <dbl>, fa_day3_rep3 <dbl>,
## #   fa_day7_rep1 <dbl>, fa_day7_rep2 <dbl>, fa_day7_rep3 <dbl>

## MA plot of 1st contrast with top 10 DE genes labelled.
plotMA(object = deseq, i = 1L, ntop = 10L)

## Volcano plot of 1st contrast with top 10 DE genes labelled.
plotVolcano(object = deseq, i = 1L, ntop = 10L)
```

**Comment by mjsteinbaugh** [comment]
Also check that this returns `TRUE` for you in an R session:

```r
options(
    "repos" = c(
        "CRAN" = "https://cloud.r-project.org",
        "AcidGenomics" = "https://r.acidgenomics.com"
    )
)
BiocManager::valid()
## TRUE
```

**Comment by mjsteinbaugh** [comment]
Hi @kokyriakidis ,

I just pushed an update to bcbioRNASeq (v0.3.41) that makes a couple of tweaks to the [R Markdown templates](https://github.com/hbc/bcbioRNASeq/tree/master/inst/rmarkdown/templates) in the package. I have [unit tests for our F1000 paper dataset](https://github.com/hbc/bcbioRNASeq/blob/master/tests/testthat-extra/test-f1000.R) that currently render correctly with the latest version of the package. That may be a useful resource for debugging. Take a look and let me know if the issue is still persisting.

Best,
Mike

**Comment by kokyriakidis** [comment]
Hi @mjsteinbaugh,

Thank you very much for following this issue. I will try to run the tremplate again and I will let you know. 

I tried to run the functional analysis template with no luck. Can you please check if it works fine for you?

Best,
Konstantinos

**Comment by mjsteinbaugh** [comment]
Yeah I ran it yesterday with the example F1000 dataset in the paper and it did work for me. Take a look at the [unit test file](https://github.com/hbc/bcbioRNASeq/blob/master/tests/testthat-extra/test-f1000.R) to see if you can get the example dataset to work on your machine as well.

**Comment by kokyriakidis** [comment]
Will do! Thanks again!

**Comment by kokyriakidis** [comment]
I can confirm that your test code you wrote in earlier comments works as intented! 
When I run:
```
options(
    "repos" = c(
        "CRAN" = "https://cloud.r-project.org",
        "AcidGenomics" = "https://r.acidgenomics.com"
    )
)
BiocManager::valid()
```
I had 13 outdated packages. This might caused the problems.

I installed bcbiornaseq through conda and when I tried to run the analysis for the first time a lot of packages needed to be installed. So I installed them manually. This may have caused some issues.

Thanks again for following this issue and helping! I really appreciate it!
